### PR TITLE
Adding E0011 explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #24407.

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
...
+struct Foo { a: u8 }
+ 
+impl ::std::ops::Index<u8> for Foo {
+    type Output = u8;
+ 
+    fn index<'a>(&'a self, idx: u8) -> &'a u8 { &self.a }
+}
+
+const a: Foo = Foo { a: 0u8};
+const b: u8 = a[0]; // Index trait is defined by the user, bad !
```
Comment (killercup) [inline src/librustc/diagnostics.rs]: All other Rust documentation I've read so far does not add spaces before exclamation marks.

Comment (steveklabnik) [inline src/librustc/diagnostics.rs]: :+1:

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: @killercup: Oh right ! <well, in french we have a space between the last word and '?'/'!'. I change that but you'll have to wait tomorrow before I rebase the second commit.

**Comment by GuillaumeGomez** [comment]
I squashed ! r+ or do you see something which might need to be improved/fixed ?

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
 the heap at runtime, and therefore cannot be done at compile time.
 "##,
 
+E0011: r##"
+Using user-defined operator on const/static variable is restricted to what can be
+evaluated at compile-time. Using an user-defined operator could call a user-defined
```
Comment (mrmonday) [inline src/librustc/diagnostics.rs]: This should be "Using a user-defined ..."

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
...
+
+struct Foo { a: u8 }
+
+impl ::std::ops::Index<u8> for Foo {
+    type Output = u8;
+
+    fn index<'a>(&'a self, idx: u8) -> &'a u8 { &self.a }
+}
+
+const a: Foo = Foo { a: 0u8};
```
Comment (mrmonday) [inline src/librustc/diagnostics.rs]: Probably want a space before the `}` here.

File: src/librustc/diagnostics.rs [hunk 4]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
...
+const b: u8 = a[0]; // Index trait is defined by the user, bad!
+```
+
+The only traits which can be used have to be already implemented, not user-defined.
+
+Example :
```
Comment (mrmonday) [inline src/librustc/diagnostics.rs]: Probably don't want the space before the `:` here.

File: src/llvm [hunk 5]
```diff
@@ -1 +0,0 @@
-Subproject commit bff69076975642c64e76dbeaa53476bfa7212086
```
Comment (mrmonday) [inline src/llvm]: Did you mean to do this?

**Comment by GuillaumeGomez** [comment]
@mrmonday: Should be good now.

File: src/librustc/diagnostics.rs [hunk 6]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
...
+    type Output = u8;
+
+    fn index<'a>(&'a self, idx: u8) -> &'a u8 { &self.a }
+}
+
+const a: Foo = Foo { a: 0u8 };
+const b: u8 = a[0]; // Index trait is defined by the user, bad!
+```
+
+The only traits which can be used have to be already implemented, not user-defined.
```
Comment (cmr) [inline src/librustc/diagnostics.rs]: I think "already implemented" is somewhat unclear, and that a better replacement would be: "only operators on builtin types are allowed."

**Comment by GuillaumeGomez** [comment]
Updated !

**Comment by Manishearth** [comment]
cc @michaelsproul

File: src/librustc/diagnostics.rs [hunk 7]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
...
+user-defined function, which is not allowed.
+
+Bad example:
+
+```
+use std::ops::Index;
+
+struct Foo { a: u8 }
+
+impl ::std::ops::Index<u8> for Foo {
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: nit: no need for a full path

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Oh right ! I totally forgot that I imported the trait just above...

File: src/librustc/diagnostics.rs [hunk 8]
```diff
@@ -176,6 +176,38 @@ for the entire lifetime of a program. Creating a boxed value allocates memory on
 the heap at runtime, and therefore cannot be done at compile time.
 "##,
 
+E0011: r##"
+Using a user-defined operator on const/static variable is restricted to what
```
Comment (michaelsproul) [inline src/librustc/diagnostics.rs]: This first sentence doesn't really make sense as you can't use user-defined operators on consts or statics at all. I think we should accentuate the point from the second sentence. Something like:

> Initializers for constants and statics are evaluated at compile time. User-defined operators rely on user-defined functions, which cannot be evaluated at compile time.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by michaelsproul** [comment]
Perfect! Ready to land! r=@Manishearth ?

**Pull request merged**